subgraphs/cbridge/src/mappings.ts

```diff
@@ -0,0 +1,1187 @@
+export function decodeWithdrawMsg(buf: Bytes): WithdrawMsg {
+  const wdmsg = new WithdrawMsg(
+    BIGINT_ZERO,
+    BIGINT_ZERO,
+    Address.zero(),
+    Address.zero(),
+    BIGINT_ZERO,
+    Bytes.empty()
+  );
+
+  let idx: i32 = 0;
+  while (idx < buf.length) {
+    const decoded = decodeKey(buf, idx);
+    const tag = decoded[0]; // attribute position in WithdrawMsg
+    const wireType = decoded[1];
+    idx = decoded[2];
+
+    log.info("[decodeWithdrawMsg]idx={},tag={},wireType={}", [
+      idx.toString(),
+      tag.toString(),
+      wireType.toString(),
+    ]);
+
+    let retBigInt: BigInt[];
+    let retBytes: Bytes[];
+    switch (tag) {
+      case 1:
+        retBigInt = decodeVarInt(buf, idx);
+        wdmsg.chainId = retBigInt[0];
+        idx = retBigInt[1].toI32();
+        break;
+      case 2:
+        retBigInt = decodeVarInt(buf, idx);
+        wdmsg.seqnum = retBigInt[0];
+        idx = retBigInt[1].toI32();
+        break;
+      case 3:
+        retBytes = decodeBytes(buf, idx)!;
+        wdmsg.receiver = Address.fromBytes(retBytes[0]);
+        idx = retBytes[1].toI32();
+        break;
+      case 4:
+        retBytes = decodeBytes(buf, idx)!;
+        wdmsg.token = Address.fromBytes(retBytes[0]);
+        idx = retBytes[1].toI32();
+        break;
+      case 5:
+        retBytes = decodeBytes(buf, idx)!;
+        wdmsg.amount = toBigInt(retBytes[0])!;
+        idx = retBytes[1].toI32();
+        break;
+      case 6:
+        retBytes = decodeBytes(buf, idx)!;
+        wdmsg.refId = retBytes[0];
+        idx = retBytes[1].toI32();
+        break;
+      default:
+        idx = skipValue(buf, idx, wireType);
+    }
+  }
+
+  return wdmsg;
+} // end decoder WithdrawMsg
+
+function decodeKey(buf: Bytes, idx: i32): i32[] {
+  const result = decodeVarInt(buf, idx);
+  const v = result[0].toI32();
+  const tag = v / 8;
+  const wiretype = v & 7;
+  return [tag, wiretype, result[1].toI32()];
+}
+
+// decode varying length integer
+function decodeVarInt(buf: Bytes, idx: i32): BigInt[] {
+  const tmpArray = new Uint8Array(10); // proto int is at most 10 bytes (7 bits can be used per byte)
+  tmpArray.set(buf.slice(idx, idx + 10)); // load 10 bytes from buf[idx, idx+10] to tmp
+  let b: i32; // store current byte content
+  let v: i32 = 0; // reset to 0 for return value
+  let endIdx: i32 = idx;
+  for (let i = 0; i < 10; i++) {
+    b = tmpArray[i]; // don't use tmp[i] because it does bound check and costs extra
+    v |= (b & 0x7f) << (i * 7);
+    if ((b & 0x80) == 0) {
+      endIdx += i + 1;
+
+      log.info("[decodeVarInt]idx={},next={},result={}", [
+        idx.toString(),
+        endIdx.toString(),
+        v.toString(),
+      ]);
+      return [BigInt.fromI32(v), BigInt.fromI32(endIdx)];
+    }
+  }
+
+  throw new Error("Invalid varint stream"); // i=10, invalid varint stream
+}
+
+// decode bytes
+function decodeBytes(buf: Bytes, idx: i32): Bytes[] | null {
```

Comment:
Can we make this generalizable and add it to the _reference_ subgraph in a separate PR for future use?

Reply:
with a method header on what it does

Reply:
Actually the byte decoding method is tied to cbridge only - I am not aware any other protocol packing data this way: https://github.com/celer-network/sgn-v2-contracts/blob/aa569f848165840bd4eec8134f753e105e36ae38/contracts/libraries/PbPool.sol#L20-L45

Reply:
ah, understood. Resolving